### ISSUE TITLE
completion_only_loss=True give wrong training steps

### ISSUE THREAD

**Issue by HuangChiEn** [huggingface/trl#4738]
### Reproduction

```python
from datasets import load_dataset, concatenate_datasets
from trl import SFTTrainer, SFTConfig

def load_all_dataset(dataset_root, format, show_used_data=False, seed=42):
    ds_lst = []
    dataset_root = dataset_root if isinstance(dataset_root, Path) else Path(dataset_root)
    for ds_path in dataset_root.glob(f"*.{format}"):
        if show_used_data:
            print(f"loading {ds_path.stem}...")
        ds = load_dataset("json", data_files=str(ds_path))["train"]
        ds_lst.append(ds)

    all_ds = concatenate_datasets(ds_lst)
    shuffled_ds = all_ds.shuffle(seed=seed)
    return shuffled_ds

training_args = SFTConfig(
        **cfg.sft_dict,
        max_grad_norm=0.3,
        warmup_ratio=0.03,
        lr_scheduler_type="linear",
        logging_steps=25,
        save_steps=250,
        save_total_limit=20,
        ddp_find_unused_parameters=False,
        completion_only_loss=True,
        gradient_checkpointing_kwargs={"use_reentrant": False},
        remove_unused_columns = False,
        label_names=["labels"],
        packing=False
    )

# len(ds) : 988828, ds[0].keys() : dict_keys(['prompt', 'completion'])
ds = load_all_dataset(...)

trainer = SFTTrainer(
        model=model,
        train_dataset=ds,
        peft_config=peft_config,
        args=training_args
    )
trainer.train()
```

<img width="484" height="128" alt="Image" src="https://github.com/user-attachments/assets/20af1dc7-49fd-4036-a4f4-fd1861dabe81" />

Batch_size = 1, gradient_accumulate=1 ; it should just give 988828 steps

exec commend : `CUDA_VISIBLE_DEVICES="0" accelerate launch finetune_completion_v4.py` or `CUDA_VISIBLE_DEVICES="0" torchrun --nproc_per_node=1 finetune_completion_v4.py`
both give wrong steps...

However, it outputs:
```
it's about 988828 x 2 = 1,977,656, so i doubt that it split my single training sample "prompt", "completion" into indepenent 2 samples ?

<img width="1765" height="37" alt="Image" src="https://github.com/user-attachments/assets/82cf76fb-a153-4a72-959b-d91a8cc46cc8" />

```

But, since the training sample is private, i can not release the reproducable data for this case.

### System Info

                                                                                                                     
- Platform: Linux-4.18.0-513.5.1.el8_9.x86_64-x86_64-with-glibc2.35                                                                     
- Python version: 3.11.11                                                                                                               
- TRL version: 0.21.0                                                                                                                   
- PyTorch version: 2.7.0                                                                                                                
- accelerator(s): NVIDIA H100 80GB HBM3, NVIDIA H100 80GB HBM3, NVIDIA H100 80GB HBM3, NVIDIA H100 80GB HBM3, NVIDIA H100 80GB HBM3, NVI
DIA H100 80GB HBM3                                                                                                                      
- Transformers version: 4.57.3- 
- Accelerate version: 1.12.0                                                                                                    [2/1976]
- Accelerate config:                                                                                                                    
  - compute_environment: LOCAL_MACHINE                                                                                                  
  - distributed_type: DEEPSPEED
  - use_cpu: False
  - debug: True
  - num_processes: 1
  - machine_rank: 0
  - num_machines: 1
  - rdzv_backend: static
  - same_network: True
  - main_training_function: main
  - enable_cpu_affinity: False
  - deepspeed_config: {'deepspeed_config_file': '/data/joseph/medlm/config/deepspeed_cfg.json', 'zero3_init_flag': True}
  - downcast_bf16: no
  - tpu_use_cluster: False
  - tpu_use_sudo: False
  - tpu_env: []
- Datasets version: 4.4.1
- HF Hub version: 0.36.0
- bitsandbytes version: not installed
- DeepSpeed version: 0.18.3
- Diffusers version: not installed 
- Liger-Kernel version: not installed
- LLM-Blender version: not installed
- OpenAI version: not installed
- PEFT version: 0.18.0

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by qgallouedec** [comment]
I cannot see the image shared.

Note that the number of steps also depends on `num_train_epoch`. In your case, it seems that you let the default value (3)

**Comment by priyanshumishra610** [comment]
Hi @HuangChiEn 👋

Thanks for the detailed report  the step-count discrepancy is interesting. My guess is that it might be related to how SFTTrainer internally handles completion_only_loss=True or processes the dataset for gradient accumulation, which could cause each "prompt", "completion" pair to be treated in a way that doubles the effective steps.

I’d be happy to help dig into this and prepare a minimal reproducible test case or a short investigation to clarify why the step count differs. Since the dataset is private, I can work with a mock dataset that mimics the same structure to try and reproduce the behavior.

Let me know if it’s okay for me to take ownership of this investigation.

**Comment by HuangChiEn** [comment]
> I cannot see the image shared.
> 
> Note that the number of steps also depends on `num_train_epoch`. In your case, it seems that you let the default value (3)

Oh...thinks for point out this issue orz...
my bad, i set epoch to 2, so it double the steps.

---

However, the steps bug may occurs in other part, let me test it & may open the other issue.... (maybe a few hours

**Comment by HuangChiEn** [comment]
> Hi [@HuangChiEn](https://github.com/HuangChiEn) 👋
> 
> Thanks for the detailed report the step-count discrepancy is interesting. My guess is that it might be related to how SFTTrainer internally handles completion_only_loss=True or processes the dataset for gradient accumulation, which could cause each "prompt", "completion" pair to be treated in a way that doubles the effective steps.
> 
> I’d be happy to help dig into this and prepare a minimal reproducible test case or a short investigation to clarify why the step count differs. Since the dataset is private, I can work with a mock dataset that mimics the same structure to try and reproduce the behavior.
> 
> Let me know if it’s okay for me to take ownership of this investigation.

TKS for your reply ~
If it cause by similar issue, i may create a minimum reproduceable code snippet. If the bug just my mistake (see above), then i'll close it ~

**Comment by priyanshumishra610** [comment]
Sounds good @HuangChiEn 👍

If you’re able to share a minimal reproducible snippet, that would be super helpful. In the meantime, I’ll also try to reproduce the behavior locally with a mock dataset to see if we can narrow down whether it’s an actual TRL behavior or something dataset-specific.

Either way, happy to help investigate 😊

**Comment by HuangChiEn** [comment]
ok, confirm it ; all program goes well, my bad ~
3ep, global bz 48 (2 gpu, grad_acc:3, bz : 8), (988828/48)*3 ~= 61803
<img width="1307" height="160" alt="Image" src="https://github.com/user-attachments/assets/d368348b-9f8b-4e8b-9ca5-f25cbbf161de" />

@priyanshumishra610 @qgallouedec 
TKS for quick reply & insightful log file analyze, issue closed.
